flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -384,57 +386,59 @@ protected Counter setupNumRecordsInCounter(StreamOperator streamOperator) {
 		}
 	}
 
-	@Override
-	public final void invoke() throws Exception {
+	private void beforeInvoke() throws Exception {
```

Comment:
👍  for splitting huge ugly methods, but I'd like to have better names for the new methods.

Reply:
Given that there is `init()` already adding something like `setup()` or `prepare()` would be confusing. Do you have something specific in mind?

Reply:
`init()` is probably the right name, but we need to ensure that all subclasses of `StreamTask` also call `super.init()` if we move the functionality to it. Similar, we could do enhance `cleanup()` to actually do all the cleanup and implementators need to invoke that first. That's symmetric to how you would implement `close()`.

Reply:
This would require changes to most of the StreamTask descendants, which are completely unrelated to the current task. I'd rather keep the scope of changes limited.

Reply:
Yes, I'd propose to prefill it in `init`, but I can see the point of not wanting to change all the subclasses.
If we cannot find a good name for the method, maybe that's a sign that the new method is actually still doing too much?
But if it's too much effort to think about it, just keep as is, so that we can finish before feature freeze.

Reply:
Yes, I'd propose to prefill it in init, but I can see the point of not wanting to change all the subclasses.
If we cannot find a good name for the method, maybe that's a sign that the new method is actually still doing too much?
But if it's too much effort to think about it, just keep as is, so that we can finish before feature freeze.